### PR TITLE
Add stream metadata

### LINKED ISSUE #1819: Add stream metadata
* new `streams.file_metadata` column - all file metadata (like ID3 tags)
* dispatch event in `Stream::createStream()` to find file metadata and populate `file_metadata` 
* handle this event for images? e.g. with exif functions

### PR DESCRIPTION
**Pull request by paolosantarsiero**
This PR introduces a new column `file_metadata` for `Streams` table and dispatch event `Stream.create` on create stream entity.

Furthermore `width` and `height` are calculated for image files using `getimagesizefromstring()`

### CODE REVIEW

File: plugins/BEdita/Core/src/Model/Entity/Stream.php [hunk 1]
```diff
@@ -195,6 +197,8 @@ protected function createStream($source)
         rewind($resource);
         $stream = new LaminasStream($resource, 'r');
 
+        $this->dispatchEvent('Stream.create', $stream);
```
Comment (stefanorosanelli) [inline plugins/BEdita/Core/src/Model/Entity/Stream.php]: Second argument has to be an array in this case
```php
$this->dispatchEvent('Stream.create', [$stream]);
```

File: plugins/BEdita/Core/src/Model/Entity/Stream.php [hunk 2]
```diff
@@ -248,4 +254,16 @@ protected function _getUrl()
 
         return $this->_properties['url'] = FilesystemRegistry::getPublicUrl($this->uri);
     }
+
+    protected function _getMetadata($source)
+    {
+        $ext_allowed = Configure::read('Streams.ext_allowed');
```
Comment (stefanorosanelli) [inline plugins/BEdita/Core/src/Model/Entity/Stream.php]: I'd keep a list of allowed mime types in a class constant for now, like `EXIF_MIME_TYPES` 
we may add a configuration later, if necessary

File: plugins/BEdita/Core/src/Model/Entity/Stream.php [hunk 3]
```diff
@@ -248,4 +254,16 @@ protected function _getUrl()
...
+        if (in_array($this->mime_type, $ext_allowed)) {
```
Comment (stefanorosanelli) [inline plugins/BEdita/Core/src/Model/Entity/Stream.php]: Return early approach is better here
```php 
if (!in_array(...)) {
  return;
}
``` 
return

File: plugins/BEdita/Core/src/Model/Entity/Stream.php [hunk 4]
```diff
@@ -248,4 +254,16 @@ protected function _getUrl()
...
+            try {
+                $this->file_metadata = json_encode(exif_read_data($source));
```
Comment (stefanorosanelli) [inline plugins/BEdita/Core/src/Model/Entity/Stream.php]: remove `json_encode` and cast to `(array)` => this way we will have a json object, not a string...

File: plugins/BEdita/Core/src/Model/Entity/Stream.php [hunk 5]
```diff
@@ -248,4 +254,16 @@ protected function _getUrl()
 
         return $this->_properties['url'] = FilesystemRegistry::getPublicUrl($this->uri);
     }
+
+    protected function _getMetadata($source)
```
Comment (stefanorosanelli) [inline plugins/BEdita/Core/src/Model/Entity/Stream.php]: avoid `virtual` getter conventions like `_get{PropName}` style here => use `readFileMetadata()` or similar

File: plugins/BEdita/Core/src/Model/Entity/Stream.php [hunk 6]
```diff
@@ -248,4 +254,16 @@ protected function _getUrl()
...
+    {
+        $ext_allowed = Configure::read('Streams.ext_allowed');
+        if (in_array($this->mime_type, $ext_allowed)) {
+            try {
+                $this->file_metadata = json_encode(exif_read_data($source));
+            } catch (\Exception $e) {
+                return null;
```
Comment (stefanorosanelli) [inline plugins/BEdita/Core/src/Model/Entity/Stream.php]: add `$this->log(...., 'warning');` to track this event as warning - use `uuid` and `filename` in message

File: plugins/BEdita/Core/src/Model/Entity/Stream.php [hunk 7]
```diff
@@ -248,4 +260,36 @@ protected function _getUrl()
 
         return $this->_properties['url'] = FilesystemRegistry::getPublicUrl($this->uri);
     }
+
+    /**
+     * Read exif data from stream
+     *
+     * @param resource $resource Resource streammed
+     * @return null
```
Comment (stefanorosanelli) [inline plugins/BEdita/Core/src/Model/Entity/Stream.php]: `@return void`

File: plugins/BEdita/Core/src/Model/Entity/Stream.php [hunk 8]
```diff
@@ -195,6 +204,9 @@ protected function createStream($source)
         rewind($resource);
         $stream = new LaminasStream($resource, 'r');
 
+        $this->readFileMetadata($source, $resource);
```
Comment (stefanorosanelli) [inline plugins/BEdita/Core/src/Model/Entity/Stream.php]: `$this->readFileMetadata($resource)`

File: plugins/BEdita/Core/src/Model/Entity/Stream.php [hunk 9]
```diff
@@ -248,4 +260,36 @@ protected function _getUrl()
...
+            rewind($resource);
+            $size = getimagesizefromstring(stream_get_contents($resource));
+
+            if ($size) {
+                $this->width = $size[0];
+                $this->height = $size[1];
+            }
+        }
+
+        if (!in_array($this->mime_type, $this::EXIF_MIME_TYPES) && function_exists('exif_read_data')) {
```
Comment (stefanorosanelli) [inline plugins/BEdita/Core/src/Model/Entity/Stream.php]: `if (!in_array($this->mime_type, $this::EXIF_MIME_TYPES) || !function_exists('exif_read_data') {`

File: plugins/BEdita/Core/tests/TestCase/Model/Entity/StreamTest.php [hunk 10]
```diff
@@ -327,4 +329,52 @@ public function testGetUrl($expected, $uuid)
...
+                "FileSize" => 5403,
+                "FileType" => 2,
+                "MimeType" => "image/jpeg",
+                "SectionsFound" => ""
+            ];
+            //remove timestamp of file only for test
+            unset($stream->file_metadata['FileDateTime']);
+
+            static::assertEquals($expected, $stream->file_metadata);
+        }
```
Comment (stefanorosanelli) [inline plugins/BEdita/Core/tests/TestCase/Model/Entity/StreamTest.php]:
```php
} else {   // `exif_read_data` not available
            static::assertEmpty($stream->file_metadata);
}
```

**Review by stefanorosanelli — APPROVED** [review]

File: plugins/BEdita/Core/tests/TestCase/Model/Entity/StreamTest.php [hunk 11]
```diff
@@ -327,4 +329,72 @@ public function testGetUrl($expected, $uuid)
...
+     * Test read exif data
+     *
+     * @covers ::readFileMetadata()
+     * @covers ::createStream()
+     */
+    public function testReadFileMetadata()
+    {
+        $path = Configure::read('Filesystem.default.path');
+        $image_test = new Stream($path . '/a4fbe302-3d5b-4774-a9df-18598def690e-image-metadata.jpeg', 'r');
+        $gif_test = new Stream($path . '/6aceb0eb-bd30-4f60-ac74-273083b921b6-bedita-logo-gray.gif', 'r');
```
Comment (batopa) [inline plugins/BEdita/Core/tests/TestCase/Model/Entity/StreamTest.php]: [minor] for convention we use camel case for variables `$imageTest`, `$gifTest`, etc...

File: plugins/BEdita/Core/tests/TestCase/Model/Entity/StreamTest.php [hunk 12]
```diff
@@ -327,4 +329,72 @@ public function testGetUrl($expected, $uuid)
...
+    /**
+     * Test failed read exif data
+     *
+     * @covers ::readFileMetadata()
+     * @covers ::createStream()
+     */
+    public function testFailedReadFileMetadata()
+    {
+        $path = Configure::read('Filesystem.default.path');
+        $gif_test = new Stream($path . '/6aceb0eb-bd30-4f60-ac74-273083b921b6-bedita-logo-gray.gif', 'r');
```
Comment (batopa) [inline plugins/BEdita/Core/tests/TestCase/Model/Entity/StreamTest.php]: [minor] `$gifTest`

**Review by batopa — APPROVED** [review]

**Pull request merged**